src/multiline/flb_ml_parser.c

```diff
@@ -340,6 +340,11 @@ void flb_ml_parser_destroy_all(struct mk_list *list)
     struct mk_list *head;
     struct flb_ml_parser *parser;
 
+    /* Ensure list is initialized */
+    if (list->next == NULL) {
```

Comment:
I think the check should not be here, if there is a mk_list_init() in place the check should not be needed.

Reply:
Additionally, that could be garbage if the list was uninitialized depending on the memory state, I don't think this check could be effective unless we modify the structure to add a canary field in the header we explicitly set when initializing / disposing of the list.

Reply:
Without this check the following NULL-pointer dereference happens (line numbers may be off by ~5 or so, due to old version):

```
=================================================================
==12==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000008 (pc 0x00000063e664 bp 0x7ffd754d2630 sp 0x7ffd754d2620 T0)
==12==The signal is caused by a READ memory access.
==12==Hint: address points to the zero page.
SCARINESS: 10 (null-deref)
    #0 0x63e664 in flb_ml_parser_destroy_all /src/fluent-bit/src/multiline/flb_ml_parser.c
    #1 0x5aac26 in flb_ml_exit /src/fluent-bit/src/multiline/flb_ml.c:1168:5
    #2 0x5ae209 in flb_parser_exit /src/fluent-bit/src/flb_parser.c:388:5
    #3 0x5825ea in flb_config_exit /src/fluent-bit/src/flb_config.c:460:5
    #4 0x581b00 in flb_config_init /src/fluent-bit/src/flb_config.c
    #5 0x56ce06 in LLVMFuzzerTestOneInput /src/fluent-bit/tests/internal/fuzzers/config_map_fuzzer.c:188:18
    #6 0x440843 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #7 0x44002a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
    #8 0x4416f9 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:757:19
    #9 0x4423c5 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:895:5
    #10 0x43172f in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #11 0x45ad82 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #12 0x7ff56ae8f082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #13 0x42216d in _start (/out/flb-it-fuzz-config_map_fuzzer_OSSFUZZ+0x42216d)
```

The problem is multiline parsers are initialised after a possible exit in the event of a failed environment initialisation (we added a condition to exit after `config->env` initialisation):

https://github.com/fluent/fluent-bit/blob/7f8f1600cbd2cace51a8eb0184f3cedc26d0339d/src/flb_config.c#L310-L320

The two options I considered to solve the above NULL-pointer deref was:
- put environment initialisation after multi-line parser initialisation
- ensure that config_exit can handle configs that have uninitialised multiline parsers.

I opted for (2) as this seems to be the most general and because it's fairly non-obvious that the ordering matters -- hence I figured there is a chance of future getting it wrong.

Reply:
I think I'd move line 314 to 307 and maybe add a comment to explain the importance of that. 

I know that grouping the initialization steps related to each sub system or component is the the nicer thing to do but in this case I saw a block of list initializations and a few of them sparated before the coroutine settings so it's not like we're ruining something nice =P 

The reason why I'd prefer to move the list initialization is that we wouldn't be depending on something implicit, there is no explanation in the code of `mk_list` or `cfl_list`  for this value to be of importance, there's no abstraction for it and it really boils down to the fact that in this case the `flb_config` instance was properly allocated using `flb_alloc` and zeroed out but that's not really a solid foundation so I'd err on the safer side.

I'm open to adding an explicit initialization field to the structure and / or a function to check the state but if we do it we should do it in both `mk_list` and `cfl_list` since the latter is supposed to take over and no new code should be written using `mk_list`.